File: LightSwordX/Socket6/TCPSocket.c
//
//  TCPSocket.c
//  LightSwordX
//
//  Created by Neko on 1/9/16.
//  Copyright © 2016 Neko. All rights reserved.
//

#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <dirent.h>
#include <netdb.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/select.h>

char* get_ip_str(const struct sockaddr *sa, char *s, unsigned int maxlen)
{
    switch(sa->sa_family) {
        case AF_INET:
            inet_ntop(AF_INET, &(((struct sockaddr_in *)sa)->sin_addr), s, maxlen);
            break;
            
        case AF_INET6:
            inet_ntop(AF_INET6, &(((struct sockaddr_in6 *)sa)->sin6_addr), s, maxlen);
            break;
            
        default:
            strncpy(s, "Unknown AF", maxlen);
            return NULL;
    }
    
    return s;
}

void fill_addrinfo_port(const struct addrinfo* addr, int port) {
    if (addr->ai_family == AF_INET) {
        struct sockaddr_in* ipv4 = (struct sockaddr_in*)(addr->ai_addr);
        ipv4->sin_port = htons(port);
    } else if (addr->ai_family == AF_INET6) {
        struct sockaddr_in6* ipv6 = (struct sockaddr_in6*)(addr->ai_addr);
        ipv6->sin6_port = htons(port);
    }
}

void tcpsocket_set_block(int socket, int on) {
    int flags = fcntl(socket, F_GETFL, 0);
    
    if (on == 0) {
        fcntl(socket, F_SETFL, flags | O_NONBLOCK);
    } else {
        flags &= ~ O_NONBLOCK;
        fcntl(socket, F_SETFL, flags);
    }
}

int tcpsocket_connect(const char* host, int port, int timeout) {
    
    struct addrinfo hints, *addrs, *p;
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    
    int s = getaddrinfo(host, NULL, &hints, &addrs);
    if (s != 0) {
        return -1;
    }

    int socketfd = -1;
    int error = -1;
    int errorlen = sizeof(error);
    
    for (p = addrs; p != NULL; p = p->ai_next) {
        
        socketfd = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
        if (socketfd == -1) {
            error = socketfd;
            continue;
        }
        
        fill_addrinfo_port(p, port);
        
        tcpsocket_set_block(socketfd, 0); // Set non-block socket
        connect(socketfd, p->ai_addr, p->ai_addrlen);
        
        fd_set fdset;
        FD_ZERO(&fdset);
        FD_SET(socketfd, &fdset);
        
        struct timeval tv;
        tv.tv_sec = timeout;
        tv.tv_usec = 0;
        
        int retval = select(socketfd + 1, NULL, &fdset, NULL, &tv);
        if (retval <= 0) {
            close(socketfd); // Error or Timeout
            socketfd = -1;
            error = retval < 0 ? -2 : -3;
            continue;
        }
        
        getsockopt(socketfd, SOL_SOCKET, SO_ERROR, &error, (socklen_t *)&errorlen);
        if (error != 0) {
            fprintf(stderr, "failed to connect: %s\n", gai_strerror(error));
            close(socketfd); // Failed to connect
            socketfd = -1;
            continue;
        }
        
        tcpsocket_set_block(socketfd, 1);
        int set = 1;
        setsockopt(socketfd, SOL_SOCKET, SO_NOSIGPIPE, (void *)&set, sizeof(int));
        
        break;
    }
    
    freeaddrinfo(addrs);
    
    return socketfd > 0 ? socketfd : error;
}

int tcpsocket_close(int socketfd) {
    return close(socketfd);
}

int tcpsocket_pull(int socketfd, char* data, int len, int timeout) {
//    if (timeout > 0) {
//        fd_set fdset;
//        FD_ZERO(&fdset);
//        FD_SET(socketfd, &fdset);
//        
//        struct timeval tv;
//        tv.tv_usec = 0;
//        tv.tv_sec = timeout;
//        
//        int ret = select(socketfd + 1, &fdset, NULL, NULL, &tv);
//        if (ret <= 0) {
//            fprintf(stderr, "select: %d \n", ret);
//            return ret;
//        }
//    }
    
    if (timeout > 0) {
        struct timeval tv;
        
        tv.tv_sec = timeout;
        tv.tv_usec = 0;
        
        setsockopt(socketfd, SOL_SOCKET, SO_RCVTIMEO, (char *)&tv,sizeof(struct timeval));
    }
    
    return (int)read(socketfd, data, len);
}

int tcpsocket_send(int socketfd, const char* data, int len) {
    int byteswrote = 0;
    while (len - byteswrote > 0) {
        int wrotelen = (int)write(socketfd, data + byteswrote, len - byteswrote);
        if (wrotelen < 0) {
            return -1;
        }
        
        byteswrote += wrotelen;
    }
    
    return byteswrote;
}

int tcpsocket_accept(int socketfd, char* remoteip, int* remoteport) {
    char clienthost[NI_MAXHOST];
    char clientservice[NI_MAXSERV];
    
    struct sockaddr_storage clientaddr;
    socklen_t addrlen = sizeof(clientaddr);
    
    int connfd = accept(socketfd, (struct sockaddr *)&clientaddr, &addrlen);
    if (getnameinfo((struct sockaddr *)&clientaddr, addrlen, clienthost, NI_MAXHOST, clientservice, NI_MAXSERV, NI_NUMERICHOST)) {
        fprintf(stderr, "can not resolve name");
    } else {
        if (clientaddr.ss_family == AF_INET) {
            struct sockaddr_in* ipv4 = (struct sockaddr_in *)&clientaddr;
            char* clientip = inet_ntoa(ipv4->sin_addr);
            memcpy(remoteip, clientip, strlen(clientip));
            *remoteport = ipv4->sin_port;
        } else if (clientaddr.ss_family == AF_INET6) {
            struct sockaddr_in6* ipv6 = (struct sockaddr_in6 *)&clientaddr;
            char clientip[INET6_ADDRSTRLEN];
            inet_ntop(AF_INET6, &(ipv6->sin6_addr), clientip, INET6_ADDRSTRLEN);
            memcpy(remoteip, clientip, strlen(clientip));
            *remoteport = ipv6->sin6_port;
        }
    }
    
    return connfd > 0 ? connfd : -1;
}

int tcpsocket_listen(const char *addr,int port){
    
    int socketfd=socket(AF_INET, SOCK_STREAM, 0);
    int reuseon = 1;
    
    setsockopt( socketfd, SOL_SOCKET, SO_REUSEADDR, &reuseon, sizeof(reuseon) );
    
    //bind
    struct sockaddr_in serv_addr;
    memset( &serv_addr, '\0', sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = inet_addr(addr);
    serv_addr.sin_port = htons(port);
    
    int r = bind(socketfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr));
    if(r == 0){
        if (listen(socketfd, 128) == 0) {
            return socketfd;
        }else{
            return -2;//listen error
        }
    }else{
        return -1;//bind error
    }
}

int tcpsocket6_listen(const char* addr, int port) {
    
    int socketfd = socket(AF_INET6, SOCK_STREAM, 0);
    
    int reuse = 1;
    setsockopt(socketfd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));
    
    if (socketfd < 0) {
        return socketfd;
    }
    
    struct sockaddr_in6 sa;
    memset(&sa, 0, sizeof(sa));
    sa.sin6_port = htons(port);
    sa.sin6_family = AF_INET6;
    sa.sin6_len = sizeof(sa);
    inet_pton(AF_INET6, addr, &(sa.sin6_addr));
    
    int b = bind(socketfd, (struct sockaddr *) &sa, sizeof(sa));
    if (b != 0) {
        fprintf(stderr, "%s\n", gai_strerror(b));
        return -1;
    }
    
    int l = listen(socketfd, 128);
    if (l !=0 ) {
        fprintf(stderr, "%s\n", gai_strerror(b));
        return -2;
    }
    
    return socketfd;
}

File: LightSwordX/TCPSocket.swift
//
//  TCPSocket.swift
//  LightSwordX
//
//  Created by Neko on 1/9/16.
//  Copyright © 2016 Neko. All rights reserved.
//

import Foundation

@asmname("tcpsocket_connect") func c_tcpsocket_connect(host:UnsafePointer<Int8>, port:Int32, timeout:Int32) -> Int32
@asmname("tcpsocket_close") func c_tcpsocket_close(fd:Int32) -> Int32
@asmname("tcpsocket_send") func c_tcpsocket_send(fd:Int32, buff:UnsafePointer<UInt8>, len:Int32) -> Int32
@asmname("tcpsocket_pull") func c_tcpsocket_pull(fd:Int32, buff:UnsafePointer<UInt8>, len:Int32, timeout:Int32) -> Int32
@asmname("tcpsocket_listen") func c_tcpsocket_listen(addr:UnsafePointer<Int8>, port:Int32)->Int32
@asmname("tcpsocket_accept") func c_tcpsocket_accept(socketfd:Int32, ip:UnsafePointer<Int8>, port:UnsafePointer<Int32>) -> Int32

class TCPClient6: Socket {
    func connect(timeout t: Int) -> (Bool, String) {
        let r = c_tcpsocket_connect(self.addr, port: Int32(self.port), timeout: Int32(t))
        
        if r > 0 {
            self.fd = r
            return (true, "connceted")
        } else {
            switch r {
            case -1:
                return (false, "query server fail")
            default:
                return (false, "unknow err")
            }
        }
    }
    
    func close() -> (Bool, String) {
        guard let fd = self.fd else {
            return (false, "socket not open")
        }
        
        c_tcpsocket_close(fd)
        self.fd = nil
        return (true, "closed")
    }
    
    func send(data d: [UInt8]) -> (Bool, String) {
        guard let fd = self.fd else {
            return (false, "socket not open")
        }
        
        let sent = c_tcpsocket_send(fd, buff: d, len: Int32(d.count))
        if Int(sent) == d.count {
            return (true, "succcess")
        } else {
            return (false, "error")
        }
    }
    
    func send(str s: String) -> (Bool, String) {
        guard let fd = self.fd else {
            return (false, "socket not open")
        }
        
        let sent = c_tcpsocket_send(fd, buff: s, len: Int32(strlen(s)))
        if sent == Int32(strlen(s)) {
            return (true, "success")
        }
        
        return (false, "error")
    }
    
    func send(data d: NSData) -> (Bool, String) {
        guard let fd = self.fd else {
            return (false, "socket not open")
        }
        
        var buf = [UInt8](count: d.length, repeatedValue: 0)
        d.getBytes(&buf, length: d.length)
        
        let sent = c_tcpsocket_send(fd, buff: buf, len: Int32(d.length))
        if sent == Int32(d.length) {
            return (true, "success")
        }
        
        return (false, "error")
    }
    
    func read(exceptLength: Int, timeout: Int = -1) -> [UInt8]? {
        guard let fd = self.fd else {
            return nil
        }
        
        var buf = [UInt8](count: exceptLength, repeatedValue: 0)
        let read = c_tcpsocket_pull(fd, buff: buf, len: Int32(exceptLength), timeout: Int32(timeout))
        if read <= 0 {
            return nil
        }
        
        return Array(buf[0...Int(read - 1)])
    }
}

class TCPServer6: Socket {
    
    func listen() -> (Bool, String) {
        let fd = c_tcpsocket_listen(self.addr, port: Int32(self.port))
        if fd > 0 {
            self.fd = fd;
            return (true, "listening")
        }
        
        return (false, "Failed to listen")
    }
    
    func accept() -> TCPClient6? {
        guard let fd = self.fd else {
            return nil
        }
        
        var buf = [Int8](count: 16, repeatedValue: 0)
        var port: Int32 = 0
        let clientfd = c_tcpsocket_accept(fd, ip: &buf, port: &port)
        if clientfd < 0 {
            return nil
        }
        
        let client = TCPClient6()
        client.fd = clientfd
        client.port = Int(port)
        if let addr = String(CString: buf, encoding: NSUTF8StringEncoding) {
            client.addr = addr
        }
        
        return client
    }
    
    func close() -> (Bool, String) {
        guard let fd = self.fd else {
            return (false, "socket not open")
        }
        
        c_tcpsocket_close(fd)
        self.fd = nil
        return (true, "closed")
    }
}